File: llm-harness-js/src/stats/image/image-stats/source-stage.ts
import { getArtificialAnalysisImageStats } from "../sources/artificial-analysis.js";
import { getArenaAiImageStats } from "../sources/arena-ai.js";

import type { ImageSourceData } from "./types.js";

export async function fetchSourceData(): Promise<ImageSourceData> {
  const [artificialAnalysisPayload, arenaPayload] = await Promise.all([
    getArtificialAnalysisImageStats(),
    getArenaAiImageStats(),
  ]);
  const artificialAnalysisModels = artificialAnalysisPayload.data ?? [];
  const arenaModels = arenaPayload.rows ?? [];

  return {
    artificialAnalysisPayload,
    arenaPayload,
    artificialAnalysisModelsBySlug: new Map(
      artificialAnalysisModels
        .filter((model) => typeof model.slug === "string")
        .map((model) => [model.slug as string, model]),
    ),
    arenaModelsByName: new Map(
      arenaModels.map((model) => [model.model, model]),
    ),
  };
}

File: llm_harness/stats/image/image_stats/source_stage.py
"""Source stage for Python image stats."""

from __future__ import annotations

from ..sources.arena_ai import get_arena_ai_image_stats
from ..sources.artificial_analysis import get_artificial_analysis_image_stats
from .types import ImageSourceData


def fetch_source_data() -> ImageSourceData:
    """Fetch source rows for Source-stage image stats selection."""
    artificial_analysis_payload = get_artificial_analysis_image_stats()
    arena_payload = get_arena_ai_image_stats()
    artificial_analysis_models = artificial_analysis_payload.get("data") or []
    arena_models = arena_payload.get("rows") or []
    return {
        "artificial_analysis_payload": artificial_analysis_payload,
        "arena_payload": arena_payload,
        "artificial_analysis_models_by_slug": {model["slug"]: model for model in artificial_analysis_models if isinstance(model, dict) and isinstance(model.get("slug"), str)},
        "arena_models_by_name": {model["model"]: model for model in arena_models if isinstance(model, dict) and isinstance(model.get("model"), str)},
    }
